File: app/src/main/java/com/uncmorfi/MainActivity.java
package com.uncmorfi;

import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.design.widget.NavigationView;
import android.support.v4.app.Fragment;
import android.support.v4.view.GravityCompat;
import android.support.v4.widget.DrawerLayout;
import android.support.v7.app.ActionBarDrawerToggle;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.view.MenuItem;

import com.google.android.gms.maps.GoogleMap;
import com.google.android.gms.maps.OnMapReadyCallback;
import com.uncmorfi.about.AboutDialog;
import com.uncmorfi.balance.BalanceFragment;
import com.uncmorfi.counter.CounterFragment;
import com.uncmorfi.help.HelpFragment;
import com.uncmorfi.map.MapFragment;
import com.uncmorfi.menu.MenuFragment;

public class MainActivity extends AppCompatActivity implements
        NavigationView.OnNavigationItemSelectedListener, OnMapReadyCallback {

    private DrawerLayout mDrawerLayout;
    private NavigationView mNavigationView;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        setToolbarAndNavigation();
        if (savedInstanceState == null) {
            setMainFragment();
        }
    }

    private void setToolbarAndNavigation() {
        Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);

        mDrawerLayout = findViewById(R.id.drawer_layout);
        ActionBarDrawerToggle toggle = new ActionBarDrawerToggle(
                this, mDrawerLayout, toolbar, R.string.navigation_open, R.string.navigation_close);
        mDrawerLayout.addDrawerListener(toggle);
        toggle.syncState();

        mNavigationView = findViewById(R.id.nav_view);
        mNavigationView.setNavigationItemSelectedListener(this);
    }

    private void setMainFragment() {
        Fragment firstFragment = new MenuFragment();
        getSupportFragmentManager()
                .beginTransaction()
                .add(R.id.content_frame, firstFragment)
                .commit();
    }

    @Override
    public boolean onNavigationItemSelected(@NonNull MenuItem item) {
        boolean changed = true;

        if (!item.isChecked())
            changed = replaceFragment(item.getItemId());

        mDrawerLayout.closeDrawer(GravityCompat.START);
        return changed;
    }

    @Override
    public void onBackPressed() {
        if (mDrawerLayout.isDrawerOpen(GravityCompat.START)) {
            mDrawerLayout.closeDrawer(GravityCompat.START);
        } else if (mNavigationView.getMenu().getItem(0).isChecked()) {
            super.onBackPressed();
        } else {
            // Go to main fragment
            replaceFragment(R.id.nav_menu);
            mNavigationView.setCheckedItem(R.id.nav_menu);
        }
    }

    private boolean replaceFragment(int item) {
        Fragment fragment = getFragmentById(item);

        if (fragment != null) {
            getSupportFragmentManager()
                    .beginTransaction()
                    .replace(R.id.content_frame, fragment)
                    .commit();
        }
        return fragment != null;
    }

    private Fragment getFragmentById(int id) {
        Fragment fragment = null;

        switch (id) {
            case R.id.nav_menu:
                fragment = new MenuFragment();
                break;
            case R.id.nav_balance:
                fragment = new BalanceFragment();
                break;
            case R.id.nav_counter:
                fragment = new CounterFragment();
                break;
            case R.id.nav_renovation:
                sendRenovationEmail();
                break;
            case R.id.nav_map:
                MapFragment mapFragment = new MapFragment();
                fragment = mapFragment;

                mapFragment.getMapAsync(this);
                break;
            case R.id.nav_help:
                fragment = new HelpFragment();
                break;
            case R.id.nav_about:
                new AboutDialog().show(getSupportFragmentManager(), "AboutDialog");
                break;
        }
        return fragment;
    }

    private void sendRenovationEmail() {
        Intent i = new Intent(Intent.ACTION_SENDTO);
        i.setData(Uri.parse("mailto:"));
        i.putExtra(Intent.EXTRA_EMAIL, new String[]{"credenciales@comedor.unc.edu.ar"});
        i.putExtra(Intent.EXTRA_SUBJECT, getString(R.string.renovation_email_subject));
        i.putExtra(Intent.EXTRA_TEXT, getString(R.string.renovation_email_body));

        if (i.resolveActivity(getPackageManager()) != null)
            startActivity(i);
    }

    @Override
    public void onMapReady(GoogleMap googleMap) {
        MapFragment fragment = (MapFragment)
                getSupportFragmentManager().findFragmentById(R.id.content_frame);

        if (fragment != null)
            fragment.onMapReady(googleMap);
    }
}

File: app/src/main/java/com/uncmorfi/MainActivity.kt
package com.uncmorfi

import android.os.Bundle
import android.util.Log
import android.view.MenuItem
import androidx.activity.viewModels
import androidx.appcompat.app.ActionBarDrawerToggle
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.GravityCompat
import androidx.core.view.WindowCompat
import androidx.fragment.app.Fragment
import com.google.android.material.navigation.NavigationView
import com.uncmorfi.ui.about.AboutDialog
import com.uncmorfi.ui.balance.BalanceFragment
import com.uncmorfi.ui.faq.FaqFragment
import com.uncmorfi.ui.home.HomeFragment
import com.uncmorfi.ui.map.MapFragment
import com.uncmorfi.ui.menu.MenuFragment
import com.uncmorfi.ui.reminders.RemindersFragment
import com.uncmorfi.ui.servings.ServingsFragment
import com.uncmorfi.shared.*
import kotlinx.android.synthetic.main.activity_main.*

class MainActivity : AppCompatActivity(), NavigationView.OnNavigationItemSelectedListener {

    private val viewModel: MainViewModel by viewModels()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        WindowCompat.setDecorFitsSystemWindows(window, false)
        setContentView(R.layout.activity_main)

        setToolbarAndNavigation()
        if (savedInstanceState == null) {
            setMainFragment()
        }

        viewModel.refreshWorkers()

        observe(viewModel.status) {
            Log.i("MainActivity", "new status: $it")
            content_layout.snack(it)
        }
    }

    private fun setToolbarAndNavigation() {
        setSupportActionBar(toolbar)

        val toggle = ActionBarDrawerToggle(
            this,
            drawerLayout,
            toolbar,
            R.string.navigation_open,
            R.string.navigation_close
        )
        drawerLayout.addDrawerListener(toggle)
        toggle.syncState()

        navView.setNavigationItemSelectedListener(this)
    }

    private fun setMainFragment() {
        supportFragmentManager
            .beginTransaction()
            .add(R.id.content_frame, HomeFragment())
            .commit()
    }

    override fun onNavigationItemSelected(item: MenuItem): Boolean {
        var changed = true

        if (!item.isChecked)
            changed = replaceFragment(item.itemId)

        drawerLayout.closeDrawer(GravityCompat.START)
        return changed
    }

    override fun onBackPressed() {
        when {
            drawerLayout.isDrawerOpen(GravityCompat.START) -> {
                drawerLayout.closeDrawer(GravityCompat.START)
            }
            navView.menu.getItem(0).isChecked -> {
                super.onBackPressed()
            }
            else -> {
                // Go to main fragment
                replaceFragment(R.id.nav_home)
                navView.setCheckedItem(R.id.nav_home)
            }
        }
    }

    private fun replaceFragment(item: Int): Boolean {
        val fragment = getFragmentById(item)

        if (fragment != null) {
            supportFragmentManager
                .beginTransaction()
                .replace(R.id.content_frame, fragment)
                .commit()
        }
        return fragment != null
    }

    private fun getFragmentById(id: Int): Fragment? {
        when (id) {
            R.id.nav_home -> return HomeFragment()
            R.id.nav_menu -> return MenuFragment()
            R.id.nav_balance -> return BalanceFragment()
            R.id.nav_servings -> return ServingsFragment()
            R.id.nav_reminders -> return RemindersFragment()
            R.id.nav_renovation -> sendEmail(
                "credenciales@estudiantiles.unc.edu.ar",
                R.string.renovation_email_subject,
                R.string.renovation_email_body
            )
            R.id.nav_map -> return MapFragment()
            R.id.nav_faq -> return FaqFragment()
            R.id.nav_face -> openFacebook()
            R.id.nav_github -> startBrowser("https://github.com/AIDEA775/UNCmorfi")
            R.id.nav_about -> AboutDialog().show(supportFragmentManager, "AboutDialog")
        }
        return null
    }

    fun change(id: Int) {
        replaceFragment(id)
        navView.setCheckedItem(id)
    }
}